In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [39]:
df = pd.read_excel('Adops & Data Scientist Sample Data.xlsx', sheet_name='Q1 Analytics')

In [40]:
df.head()

ts user_id country_id site_id
0  2019-02-01 00:01:24  LC36FC        TL6   N0OTG
1  2019-02-01 00:10:19  LC39B6        TL6   N0OTG
2  2019-02-01 00:21:50  LC3500        TL6   N0OTG
3  2019-02-01 00:22:50  LC374F        TL6   N0OTG
4  2019-02-01 00:23:44  LCC1C3        TL6   QGO3G

Q1. Consider only the rows with country_id = "BDV" (there are 844 such rows). For each site_id, we can compute the number of unique user_id's found in these 844 rows. Which site_id has the largest number of unique users? And what's the number?

In [111]:
df_BDV = df[df['country_id']=="BDV"]
df_BDV.groupby('site_id').nunique()['user_id'].sort_values(ascending=False)

site_id
5NPAU    544
N0OTG     90
3POLC      2
Name: user_id, dtype: int64

5NPAU has the largest number of unique users. There are 544 of unique user_id.

Q2. Between 2019-02-03 00:00:00 and 2019-02-04 23:59:59, there are four users who visited a certain site more than 10 times. Find these four users & which sites they (each) visited more than 10 times. (Simply provides four triples in the form (user_id, site_id, number of visits) in the box below.)

First, I transformed our time_stamp column in to date type and get the range of the time.

In [42]:
df['ts'] = pd.to_datetime(df['ts'])  
mask = (df['ts'] > '2019-02-03 00:00:00') & (df['ts'] <= '2019-02-04 23:59:59')

In [118]:
df_question2 = df[mask].groupby(['user_id', 'site_id']).agg('count')
df_question2[df_question2['country_id'] > 10]['country_id'].sort_values(ascending=False)

user_id  site_id
LC3A59   N0OTG      26
LC06C3   N0OTG      25
LC3C9D   N0OTG      17
LC3C7E   3POLC      15
Name: country_id, dtype: int64

Q3. For each site, compute the unique number of users whose last visit (found in the original data set) was to that site. For instance, user "LC3561"'s last visit is to "N0OTG" based on timestamp data. Based on this measure, what are top three sites? (hint: site "3POLC" is ranked at 5th with 28 users whose last visit in the data set was to 3POLC; simply provide three pairs in the form (site_id, number of users).)

In [115]:
df_max = df.iloc[df.groupby('user_id')['ts'].idxmax()][['user_id', 'site_id']]
df_max.groupby('site_id').agg('count').sort_values(by='user_id', ascending=False)

user_id
site_id         
5NPAU        992
N0OTG        561
QGO3G        289
GVOFK         42
3POLC         28
RT9Z6          2
EUZ/Q          1
JSUUP          1

The df_max data frame will record the user_id and their last visited site_id. Therefore, using the groupby method and aggregate function will get the top three sites: 5NPAU, N0OTG, QGO3G.

Q4. For each user, determine the first site he/she visited and the last site he/she visited based on the timestamp data. Compute the number of users whose first/last visits are to the same website. What is the number?

In [116]:
df_min = df.iloc[df.groupby('user_id')['ts'].idxmin()][['user_id', 'site_id']]
sum(df_max['site_id'].values == df_min['site_id'].values)

1670

This question is similiar to the previous one. We get the first visited site of each user_id and compare it to the last visited site.